### ISSUE TITLE
Gather docs about internal design of workflow and workflow UI and publish them somewhere

### ISSUE THREAD

**Issue by zach-klippenstein** [square/workflow-kotlin#464]
We've published some documentation on how to _use_ workflow, but not how it works under the hood.

**Comment by zach-klippenstein** [comment]
I think we have a lot of design-doc style material scattered across multiple issues and PRs over the last few years. It's there if you dig, but not discoverable _at all_.

**Comment by lucamtudor** [comment]
I've seen a couple of issues that discuss API designs, but they're hard to follow since I don't have all the context around them. It would be great to see some material in one place, or at least a list of issues to read in the correct order.
